### PR TITLE
Extract provider_job_id at execution time for delayed_job adapter

### PR DESCRIPTION
**Pull request by rchekaluk**
### Summary
Enhance delayed_job adapter to provide `provider_job_id` at execution time as mentioned in https://github.com/rails/rails/pull/26587#issuecomment-249314688. This change scans persisted delayed_job jobs to find the matching Active Job, and returns the associated DelayedJob id from it.

Analogous to https://github.com/rails/rails/pull/25961 for sidekiq.

### CODE REVIEW

**Comment by rchekaluk** [comment]
All checks have passed
This branch has no conflicts with the base branch

**Comment by rafaelfranca** [comment]
Thank you for the pull request but I'm not sure if it worth the price of going to the database at execution time to grab this id. Can you explain why you need it?

**Comment by rchekaluk** [comment]
I don't necessarily require this feature; I have been able to get by without it.

However, I offer this PR primarily for general feature completeness: expanding the support for provider_job_id at execution time (delayed_job adapter). In the same spirit as https://github.com/rails/rails/issues/26581 and https://github.com/rails/rails/pull/25961, it's *counterintuitive for provider_job_id to have one value at enqueue time,  and a different/missing value at execution time*. Therefore, given that provider_job_id is exposed, this PR seeks to remedy this confusion by returning a consistent value at all times.

When managing delayed jobs, it's plausible to imagine that developers might need to "drop down" into the delayed_job layer and access the delayed job itself (analogous to [using tracking-progress methods in Sidekiq]( https://github.com/rails/rails/pull/25961#issue-78933306)):
```
class ApplicationJob < ActiveJob::Base

  def delayed_job
    begin
      dj = Delayed::Job.find(provider_job_id)
    rescue ActiveRecord::RecordNotFound => e
      return nil
    end

    return dj
  end

end
```

It may also be worthwhile to explicitly document the behaviors of provider_job_id (in [API docs](https://api.rubyonrails.org/classes/ActiveJob/Base.html)? In [Active Job Basics](https://guides.rubyonrails.org/active_job_basics.html)?), and use this to highlight any performance concerns. Something like this:

| Adapter | after_enqueue | before_perform |
| ------------- | ------------- | ------------- |
| async | supported  | nil *(pull requests accepted)* |
| backburner | supported  | nil *(pull requests accepted)* |
| delayed_job | supported  | supported for ActiveRecord backend<br>*(Performance note: each call scans the backend persistence)* |
| inline | supported  | nil *(pull requests accepted)* |
| qu | supported  | nil *(pull requests accepted)* |
| que | supported  | nil *(pull requests accepted)* |
| queue_classic | supported  | nil *(pull requests accepted)* |
| resque | supported  | nil *(pull requests accepted)* |
| sidekiq | supported  | supported |
| sneakers | supported  | nil *(pull requests accepted)* |
| sucker_punch | supported  | nil *(pull requests accepted)* |

**Comment by rafaelfranca** [comment]
The problem with the way it is right now is that it doesn't allow people to opt-out the overhead if they don't need that information. I'm more inclined to do the SQL when the user try to access the `provider_job_id`.

**Comment by rchekaluk** [comment]
I've pushed an alternate implementation that defers determining `provider_job_id` until it is needed. Is this what you had in mind?

Note: I believe that monkey-patching the `provider_job_id` method does not impact other adapters because only one adapter may be used at a time.

**Comment by rafaelfranca** [comment]
Yes, this is what I had in mind. But instead of monkey patch we can just implement the logic in the adapters and call the adapter from `Core`.

**Comment by rchekaluk** [comment]
Done

File: activejob/lib/active_job/queue_adapters/delayed_job_adapter.rb [hunk 1]
```diff
@@ -27,6 +27,24 @@ def enqueue_at(job, timestamp) #:nodoc:
...
+          job_id_aj = job.job_id
+          djs = scan_backend(job_id_aj)
+          djs.map do |dj|
+            obj = dj.payload_object
+            next if obj.blank? || obj.job_data.blank?
+
+            job_id_persisted = obj.job_data["job_id"]
+            return dj.id if job_id_persisted == job_id_aj
+          end
+        rescue
```
Comment (rafaelfranca) [inline activejob/lib/active_job/queue_adapters/delayed_job_adapter.rb]: Why this rescue here?

Comment (rchekaluk) [reply activejob/lib/active_job/queue_adapters/delayed_job_adapter.rb]: In the original commit, I suspect it was primarily intended to rescue the case that the backend (in that commit, ActiveRecord only) raises an exception when no records are found.

However, now that the backend access has been abstracted into its own method, the rescue should be moved to that method (even though (1) ActiveRecord appears to only return an empty collection when no records are found, and (2) ActiveRecord lazy queries are not executed within that method; perhaps future supported backends will raise such an exception)

I have committed a change that moves the rescue, although I will defer to your judgement.

**Comment by rchekaluk** [comment]
Ping @rafaelfranca ?

**Pull request closed without merging**